### PR TITLE
Plugin mechanism

### PR DESCRIPTION
**Pull request by tkurki**
A plugin mechanism should allow extending the server's functionalities. Some examples:
- derived data: calculate value X based on some other values and include the derived data in the SK outputs
- conversion and output to 0183/N2K:
  - create 0183 autopilot data and send it to connected tillerpilot
  - convert AIS data from connected 0183 ais receiver to N2K
- integration to external services: for example send location to MarineTraffic
- logging: write all numeric data to 

A plugin needs access to
- SK input stream and full model data
- 0183 and N2K input streams
- SK output stream & full model
- all outputs (I2C, 0183, N2K, etc)
- configuration data persistence
- configuration data editing

Plugins should be installed independently, allowing independent versioning & dependencies.

There should be a standard plugin configuration UI.

### CODE REVIEW

**Comment by tkurki** [comment]
I've been thinking that configuration could be based on JSON schema and the UI & data model would be dynamically generated.

The server would get a JSON schema for the configuration UI via plugin API and create the UI using for example https://github.com/mozilla-services/react-jsonschema-form. The server would take care of saving & retrieving the configuration data and plugin lifecycle.

**Comment by tkurki** [comment]
My current idea is that plugins would be independent npm modules, installed as usual with `npm install`, but with a certain keyword in `package.json` identifying them as SK Node server plugins. The server can list the installed plugins based on this keyword.

**Comment by tkurki** [comment]
The configuration UI could be browser-only, with no server side html generation apart from initial bootstrap. 

The UI would use server's REST api to retrieve & store per plugin configuration data. This part of the architecture could be part of the shared SK spec.

**Comment by tkurki** [comment]
During sailing season 2016 I had the following server side pluginish functionalities
- raw data logging (hourly files) with sync to S3
- SK data to pgn 129039 (AIS dynamic data)
- SK heading to pgn 127250
- SK data to MWV, RMC, APB
- SK data to InfluxDB
- SK data to MarineTraffic position report (AIS)

My plan is to turn these into plugins and eat this PRs dogfood with them.

**Comment by tkurki** [comment]
#102 could be another plugin.

**Comment by tkurki** [comment]
Capability to intercept and capture the data the server is passing out would be useful. 

For example primary + backup: send data from backup source only if data from primary source is not available.

**Comment by tkurki** [comment]
- [ ] add server's index and demote bower plugins one level down
- [ ] documentation
- [ ] server-driven enable/disable mechanism

**Comment by joabakk** [comment]
This is a really good initiative and should be able to cover many needs/requests. With good documentation and examples how to develop own plugins, we could have ourself an app store.

**Comment by tkurki** [comment]
@joabakk  if you have the time it would be really great if you could try the plugins branch and possibly try to create your own plugin. See one of the plugins added in 91f0ab59b37869bc739498770cc245c1c1651d3f for examples.

**Comment by joabakk** [comment]
I'm looking into it. But as far as I can see, none of these plugins are published on npm and npm update does not add these to my server. Would it be possible to make another example that takes sk data, does some calculation and sends back sk to the server? It could well be empty for me to fill in the paths. 

Ideas for plugins:

- Calculate propeller slip

- Look up polar and target speeds from db

- Generate seatalk datagrams to be sent as $STALK sentences

**Comment by tkurki** [comment]
Ready for merging - I would appreciate any and all feedback. In case somebody bites: start from Readme.

File: plugin-config/public/index.html [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  </head>
+  <body>
+    <div class="container">
+    <h1>Signal K Server Plugin Configuration</h1>
+
+    <div id="main"></div>
+    </div>
+
+    <!-- jQuery (necessary for Bootstrap's JavaScript plugins) -->
+    <script src="/bower_components/jquery/dist/jquery.min.js"</script>
```
Comment (joabakk) [inline plugin-config/public/index.html]: missing >

File: plugin-config/public/index.html [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    <!-- Include all compiled plugins (below), or include individual files as needed -->
+    <script src="/bower_components/bootstrap/dist/js/bootstrap.min.js"></script>
+
+    <script src="main.js"></script>
+
```
Comment (joabakk) [inline plugin-config/public/index.html:34]: main.js is found in developer tools, but not on the webpage

**Comment by tkurki** [comment]
I believe this is ready for merging.

**Comment by sbender9** [comment]
It would be nice to be able to do a quick check to see if a specific plugin is installed and enabled. I currently have to get the config for all of the plugins and go through them.

**Comment by tkurki** [comment]
@sbender9 you mean via http / rest?

**Comment by tkurki** [comment]
See 43685884ac385e1f61a069dd101bcc36707f21cc

**Comment by sbender9** [comment]
Yep. Perfect. Thanks!

**Pull request merged**